src/evo/specialtx.cpp

```diff
@@ -94,13 +120,30 @@ bool ProcessSpecialTxsInBlock(const CBlock& block, const CBlockIndex* pindex, CV
             return false;
         }
     }
-    // !TODO: Process batch of special txes in deterministic manager
+
+    if (!deterministicMNManager->ProcessBlock(block, pindex, state, fJustCheck)) {
+        // pass the state returned by the function above
+        return false;
+    }
+
+    // !TODO: ProcessBlock llmq quorum block processor
     return true;
 }
 
-bool UndoSpecialTxsInBlock(const CBlock& block, const CBlockIndex* pindexPrev)
+bool UndoSpecialTxsInBlock(const CBlock& block, const CBlockIndex* pindex)
 {
-    /* undo special txes in batches */
+    if (!deterministicMNManager->UndoBlock(block, pindex)) {
+        return false;
+    }
+    // !TODO: UndoBlock llmq quorum block processor
     return true;
 }
 
+uint256 CalcTxInputsHash(const CTransaction& tx)
+{
+    CHashWriter hw(CLIENT_VERSION, SER_GETHASH);
+    for (const auto& in : tx.vin) {
+        hw << in.prevout;
```

Comment:
Had an idea about this, we need to add the shield spends to the inputs hash too.

We are not protected for replay attacks if the payload is in a shield spend transaction. As the inputs vector is empty, the hash will be always the same (null). So an attacker can use the same payload on different transactions.

Reply:
Yep, I was planning to update this when adding the ability to create special txes with shield inputs (right now is not possible, not even in #2267, because it relies on FundTransaction, which is not shield-enabled).

But yeah, better to prepare the validation here, so we don't forget later.

Reply:
added in 17125b99ede30bf8180d2031e1734b36df32a301